### ISSUE TITLE
Make creationTimestamp a pointer

### ISSUE THREAD

**Issue by nikhita** [kubernetes/kubernetes#67610]
Context: https://github.com/kubernetes/kubernetes/pull/67562#discussion_r211267314.

`creationTimestamp` is a struct (metav1.Time) right now:

https://github.com/kubernetes/kubernetes/blob/6d51311735705ecf30025df91216a82f397c9cfb/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L175

However, it also has the `omitempty` json tag. The current behaviour at various places is to remove the field if it is an empty value.

From the [official docs](https://golang.org/pkg/encoding/json/#Marshal),

>  The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string. 

This means that `omitempty` holds no meaning for structs. No matter what value `creationTimestamp` has, it should persist in the object. To truly adapt to it's intended behaviour, it should be a pointer (*time.Time), like `deletionTimestamp`. It also makes sense to have `creationTimestamp` as optional logically.

Right now, we don't allow an empty `creationTimestamp` to persist anyway, so it doesn't create a problem per se.

/sig api-machinery
/cc @sttts @liggitt @deads2k @lavalamp

**Comment by lavalamp** [comment]
I agree this would be more consistent, but not sure it's a priority. Can you make an object that doesn't have this set?

**Comment by lavalamp** [comment]
@cheftako suggests that maybe this field really ought to be blank when objects are sent to the POST admission webhooks, as they haven't been created yet...

**Comment by liggitt** [comment]
> @cheftako suggests that maybe this field really ought to be blank when objects are sent to the POST admission webhooks, as they haven't been created yet...

not high priority, but results in ugly things in audit (request body of create requests), webhook payloads for create requests, and exported objects (always serialize with a null field), etc. if we can fix it, I'd take a fix.

**Comment by liggitt** [comment]
the main difficulties would be go compatibility:
* change the metav1.Object interface GetCreationTimestamp/SetCreationTimestamp methods to use pointers or change implementations to treat missing attributes as zero-values?
* ensure all places that use CreationTimestamp handle nil cases

as long as null values are omitted during serialization, wire compatibility should not be an issue

**Comment by liggitt** [comment]
xref https://github.com/golang/go/issues/11939

**Comment by lavalamp** [comment]
Another option would be to implement the Marshaller interface in metav1.ObjectMeta...

**Comment by lavalamp** [comment]
After investigating this for longer than I care to admit, I think a golang fix is probably the best thing.

Switching to a pointer risks NPE in all existing clients since an existing function call would not cause a compile error.

Implementing Marshaler in the enclosing object doesn't help people using alternative json encoders, and is extremely inconvenient. It might add an unacceptable slow down, I didn't measure.

**Comment by liggitt** [comment]
> Another option would be to implement the Marshaller interface in metav1.ObjectMeta...

When I tried that, it broke the stdlib json encoder as well because ObjectMeta is inlined everywhere, so it made the top-level object implement json.Marshaler and just encode metadata.

**Comment by lavalamp** [comment]
> When I tried that, it broke the stdlib json encoder as well because ObjectMeta is inlined everywhere, so it made the top-level object implement json.Marshaler and just encode metadata.

Yeah, that sounds like another bug, for which a fix could change behavior for a bunch of existing uses. :(

**Comment by bgrant0607** [comment]
This is still necessary.
```
#!/bin/bash
[#kube](https://twitter.com/hashtag/kube?src=hashtag_click)-gen type args
res="${1}"
shift 1
if [[ "${res}" != namespace ]] ; then
  namespace="--namespace=example"
else
  namespace=""
fi
kubectl create "${res}" -o yaml --dry-run=client "${@}" ${namespace} |\
egrep -v "creationTimestamp|status" | tee ${res}.yaml
```

**Comment by yhrn** [comment]
This also breaks server-side apply idempotency. I'm implementing an operator using controller-runtime and the [SSA client code there](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/client/patch.go#L207) uses standard JSON marshalling of the object, which results in `"creationTimestamp": null`, and this causes the API server to increment `resourceVersion` even if nothing else changes. The only workaround I have found is to fetch the object and grab `creationTimestamp` from there, but that partially defeats the purpose of SSA.

I feel like this is pretty bad since SSA should be a good fit for many operator controller use case but because of this problem it's partially broken if you use the standard libraries.

**Comment by liggitt** [comment]
> the SSA client code there uses standard JSON marshalling of the object

that's not the correct way to make a SSA patch... it includes zero-values for too many fields, not just creationTimestamp

https://github.com/kubernetes/client-go/tree/master/applyconfigurations is the library to use for making SSA patches for standard types

**Comment by liggitt** [comment]
as an example, https://github.com/kubernetes/client-go/blob/master/applyconfigurations/apps/v1/deployment.go#L39-L63 can extract fields from a Deployment object to make an SSA patch

**Comment by yhrn** [comment]
> > the SSA client code there uses standard JSON marshalling of the object
> 
> that's not the correct way to make a SSA patch... it includes zero-values for too many fields, not just creationTimestamp
> 
> https://github.com/kubernetes/client-go/tree/master/applyconfigurations is the library to use for making SSA patches for standard types

Do you mean that for every custom type that I'd want to use SSA with I should implement something similar to that or do you mean that controller-runtime is doing it wrong?

**Comment by nickpatton** [comment]
This issue is impacting a project I'm working on where we're programmatically creating Argo Rollout resources:

```
	rollout := rolloutsv1a1.Rollout{
		TypeMeta: v1.TypeMeta{
			Kind:       "Rollout",
			APIVersion: "argoproj.io/v1alpha1",
		},
		ObjectMeta: v1.ObjectMeta{
			Name:      applicationName,
			Namespace: namespace,
			Labels:    labels,
		},
		Spec: rolloutsv1a1.RolloutSpec{
			Replicas:             &replicas,
			RevisionHistoryLimit: &defaultRevisionHistory,
			Selector: &v1.LabelSelector{
				MatchLabels: labels,
			},
			Strategy: rolloutsv1a1.RolloutStrategy{
				BlueGreen: &rolloutsv1a1.BlueGreenStrategy{
					ActiveService:         applicationName + "-svc",
					PreviewService:        applicationName + "-preview-svc",
					AutoPromotionEnabled:  &helperTrue,
					ScaleDownDelaySeconds: &scaleDownDelaySeconds,
				},
			},
			Template: corev1.PodTemplateSpec{
				ObjectMeta: v1.ObjectMeta{
					Labels: labels,
				},
				Spec: corev1.PodSpec{
					Containers: []corev1.Container{
						{
							Image:           image,
							ImagePullPolicy: corev1.PullIfNotPresent,
							Name:            applicationName,
							Ports: []corev1.ContainerPort{
								{
									Name:          "app",
									ContainerPort: 5000,
								},
							},
						},
					},
					RestartPolicy: "Always",
				},
			},
		},
	}
```
The generated YAML manifest from this object is causing this error when it's applied into a Kubernetes cluster running v1.26:

`Error from server (BadRequest): error when creating "manifest.yaml": Rollout in version "v1alpha1" cannot be handled as a Rollout: strict decoding error: unknown field "spec.template.metadata.creationTimestamp"`

Currently, we've gotten around it by managing our own copy of the Rollout CRD with creationTimestamp added as a field to spec.template.metadata. It'd be great if `creationTimestamp: null` didn't show up at all in the generated Rollout object though. Should CRDs be fixed to include the creationTimestamp field as a property of metadata or is there still a chance that creationTimestamp would be changed to a pointer?

**Comment by kerthcet** [comment]
Had a terrible day for this :(

I created a crd just looks like:
```
type ServiceSpec struct {
	WorkloadTemplate DeploymentSpec `json:"workloadTemplate"`
}
```

When generating ApplyConfigurations, I'll get one like:
```
type ServiceSpecApplyConfiguration struct {
	WorkloadTemplate  *DeploymentSpec `json:"workloadTemplate,omitempty"`
}
```

So here's the problem, the `DeploymentSpec` includes a `PodTemplateSpec`, but there's no ApplyConfigurations generated for podTemplateSpec, so it will always have a null `creationTimestamp`, however, this is not part of the schema, so when patching, it always emit error: `metadata.creationTimestamp: field not declared in schema"`

`Create` OP will pass, but I need to use patch method, I have tried ways I can think of to solve this, but didn't work. Would like to seek for some help.

**Comment by kerthcet** [comment]
Played a trick to solve this problem by implementing a new converter and skipped the creationTimestamp in parsing.

**Comment by liggitt** [comment]
it looks like https://github.com/golang/go/issues/45669#issuecomment-2316498707 is trending towards getting accepted, which means we could put a json omitzero tag on the creationTimestamp field without switching it to a pointer and avoid the null serialization 🎉

**Comment by dims** [comment]
>  I think we could put a json omitzero tag on the creationTimestamp field without switching it to a pointer and avoid the null serialization 🎉

sweet!

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/130989 adds omitzero support and updates ObjectMeta creationTimestamp to omitzero, so 1.34+ will no longer serialize `creationTimestamp: null` 🎉

**Comment by bgrant0607** [comment]
Thank you @liggitt !
